DEV-44293 - Support Logz Datasource

yasmin-tr opened the pull request:
**What is this feature?**

Support the Logz.io query datasources by adding necessary additions into the elasticsearch and prometheus calls and throughout the system.
Changes were added on the entry points of the different query APIs, to pass the headers of the requests, and in the client side, where the requests to the datasource is being sent.


**Why do we need this feature?**

To be able to query datasources that are managed in logzio we need to pass relevant headers and params so the query services APIs will accept the calls.



** PR Changes **:

* Added required changes in elasticsearch and prometheus clients to add logz headers so we can call logz datasources
* Refactored implementation in logz alert evaluation so we only pass headers because url change is no longer required
* Added fix to alert evaluation to check if rule routine exists instead of getOrCreate which caused routines not to be created via scheduler if eval called before scheduler updates registry.
* Added to custom.ini required configs for logz alert evaluation to work (we will probably remove in the future 

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

yasmin-tr committed:
DEV-44293 - Support Logz Datasource
* Added required changes in elasticsearch and prometheus clients to add logz headers so we can call logz datasources
* Refactored implementation in logz alert evaluation so we only pass headers because url change is no longer required
* Added fix to alert evaluation to check if rule routine exists instead of getOrCreate which caused routines not to be created via scheduler if eval called before scheduler updates registry.
* added to custom.ini required configs for logz alert evaluation to work

ohadza commented on pkg/models/headers_logzio.go:
```diff
@@ -11,6 +11,8 @@ type LogzIoHeaders struct {
 var logzioHeadersWhitelist = []string{
 	"user-context",
+	"X-Logz-Query-Context",
+	"Query-Source",
```
depend on the relevant comment in [m3-query-service](https://github.com/logzio/gaia-full/pull/14830)

yasmin-tr commented:
run-test

yasmin-tr committed:
DEV-44293 - Support Logz Datasource
* Added relevant header for m3 query ervice endpoints on prometheus client
* Added on elasticsearch the relevant queryParams for querySource

yasmin-tr commented:
run-test

yasmin-tr committed:
Merge branch 'v10.4.x-logzio' into DEV-44293-support-logz-datasources

yasmin-tr commented:
run-test

yasmin-tr committed:
added comment

yasmin-tr commented:
run-test

yasmin-tr commented:
run-test

npololnskii commented:
run-test

yasmin-tr committed:
Merge branch 'v10.4.x-logzio' into DEV-44293-support-logz-datasources

yasmin-tr commented:
run-test

yasmin-tr committed:
Fix user-context header to be unescaped
Fix for prometheus_test.go to work with changes
Logz comments in missing places

yasmin-tr commented:
run-test

yasmin-tr committed:
Fix schedule_unit_test.go
